### PR TITLE
Add extended open multilingual wordnet reader

### PR DESCRIPTION
**Pull request by ExplorerFreda**
Related issue: https://github.com/nltk/nltk/issues/1611 
Related NLTK data issue: https://github.com/nltk/nltk_data/issues/179
Extended open multilingual wordnet: http://compling.hss.ntu.edu.sg/omw/summx.html

### CODE REVIEW

**Comment by tomaarsen** [comment]
Hey! We use [`pre-commit`](https://pre-commit.com/) hooks to ensure code quality through our [`pre-commit config`](https://github.com/nltk/nltk/blob/develop/.pre-commit-config.yaml), which runs `pyupgrade`, `black`, `isort` and some other hooks. Using it is as simple as `pip install pre-commit` and then running `pre-commit install`. From then onwards, any time you commit to `https://github.com/ExplorerFreda/nltk`, it will run these hooks before committing. 
The first time you might not have any changes to turn into a commit, so you can do `pre-commit run --all` or try to create an empty commit, and it will format your changes. You can then push this to make the CI stop complaining.

File: nltk/corpus/reader/extended_omw.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+For details about Open Multilingual WordNet, see:
+http://compling.hss.ntu.edu.sg/omw/
+
+For details about Extended Open Multilingual WordNet, see:
+http://compling.hss.ntu.edu.sg/omw/summx.html
+"""
+
+import os
+from collections import defaultdict
+from IPython.core.pylabtools import configure_inline_support
```
Comment (tomaarsen) [inline nltk/corpus/reader/extended_omw.py]: This seems like an unused import. Should this be there?

File: nltk/corpus/reader/extended_omw.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+from traitlets.traitlets import default
```
Comment (tomaarsen) [inline nltk/corpus/reader/extended_omw.py]: This seems like an unused import. Should this be there?

File: nltk/corpus/reader/extended_omw.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
+# Natural Language Toolkit: Extened Open Multilingual WordNet Reader
```
Comment (tomaarsen) [inline nltk/corpus/reader/extended_omw.py]:
```suggestion
# Natural Language Toolkit: Extended Open Multilingual WordNet Reader
```

File: nltk/corpus/reader/extended_omw.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+    def _load_lang_data(self, lang):
+        """Load the wordnet data to self._lang_data"""
+        if lang in self._lang_data.keys():
+            return
+        # load open multilingual wordnet
+        for prov in self.provenances[lang]:
+            with self._omw_reader.open(f"{prov}/wn-data-{lang}.tab") as fp:
+                self.custom_lemmas(fp, lang)
+                fp.close()
```
Comment (tomaarsen) [inline nltk/corpus/reader/extended_omw.py]: The `with` keyword already automatically closes any files whenever the `with` scope is exited, so this line is unnecessary.

File: nltk/corpus/reader/extended_omw.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        # load extended open multilingual wordnet 
+        for prov in self.exomw_provenances[lang]:
+            with self._exomw_reader.open(f"{prov}/wn-{prov}-{lang}.tab") as fp:
+                self.custom_lemmas(fp, lang)
+                fp.close()
```
Comment (tomaarsen) [inline nltk/corpus/reader/extended_omw.py]: Idem.

File: nltk/corpus/reader/extended_omw.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+                if len(pair) == 1 or pair[0] == lang:
+                    if attr == "lemma":
+                        val = val.strip().replace(" ", "_").lower()
+                        self._lang_data[lang][1][val].add(offset_pos)
+                    if attr in self.lg_attrs:
+                        index = self.lg_attrs.index(attr)
+                        self._lang_data[lang][index][offset_pos].add(val)
+
+    def disable_custom_lemmas(self, lang):
+        """prevent synsets from mistakenly added"""
```
Comment (tomaarsen) [inline nltk/corpus/reader/extended_omw.py]: Is this missing "being"?

**Review by tomaarsen — CHANGES_REQUESTED** [review]
I recognise that this is a bit of an annoying situation, because any (automated) tests will fail due to `extended_omw` not yet existing in `nltk_data`, but perhaps you can provide a bit of sample code that should work if we (manually) download `extended_omw` and place it in our `nltk_data` folder?

Beyond that I added some small nitpicks. I can't really comment on the "real functionality", because I haven't been able to run it.

**Comment by ExplorerFreda** [comment]
> I recognise that this is a bit of an annoying situation, because any (automated) tests will fail due to `extended_omw` not yet existing in `nltk_data`, but perhaps you can provide a bit of sample code that should work if we (manually) download `extended_omw` and place it in our `nltk_data` folder?
> 
> Beyond that I added some small nitpicks. I can't really comment on the "real functionality", because I haven't been able to run it.

Hi @tomaarsen, thanks for the review and suggestions! I've changed the code correspondingly. 

Below is the script for downloading the required data that should work: 
```bash
cd $NLTK_DATA_DIR/corpora/
wget http://compling.hss.ntu.edu.sg/omw/wn-wikt.zip -O extended_omw.zip
unzip extended_omw.zip
mv data/ extended_omw/
```

File: nltk/corpus/reader/extended_omw.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
+# Natural Language Toolkit: Exteneded Open Multilingual WordNet Reader
```
Comment (tomaarsen) [inline nltk/corpus/reader/extended_omw.py]:
```suggestion
# Natural Language Toolkit: Extended Open Multilingual WordNet Reader
```
I haven't had the time to take a full look at this PR yet, but I noticed this small typo.

**Comment by ekaf** [comment]
It would be nice to add the following capabilities:

1. Loading the English wiktionary wordnet. This requires small modifications in the wordnet module because it does not anticipate English wordnets from alternative provenances at this moment.

2. Loading different wordnets for the same language separetely would allow to compare the quality of different provenances. The reason why the 'wikt' and 'cldr' provenances did not go into the official OMW 1.4 package was that their quality was deemed uncertain. For each language, this PR uses the set union of all provenances, which may be appropriate for some uses, but it would be nice to also support different use cases.

3. Combining extended_wordnet with newer wordnet packages like wordnet31 and wordnet2021. This would be easier if the extended_omw corpus reader was a leaner class, that could be added to any English Wordnet, including the forthcoming ones like Open English Wordnet 2022.

4. For completeness, maybe also consider adding back the Persian wordnet and the alternative Chinese wordnet from the old OMW package (in the 'pas' and 'qcn' directories), which were left out of OMW 1.4 for quality reasons though it is possible that some users miss them.

**Comment by stevenbird** [comment]
Resolved in #2946

**Comment by stevenbird** [comment]
@ExplorerFreda thanks for your work, which is appreciated even though we ended up accepting a different PR.

**Pull request closed without merging**